File: BlackjackApp/Domain/Engine/HandEvaluator.swift
import Foundation

struct EvaluationResult {
    let outcome: HandOutcome
    let payoutMultiplier: Double  // 1.5 = BJ, 1 = win, 0 = push, -1 = loss/bust
}

enum HandEvaluator {

    // MARK: – Win/Loss/Push

    static func outcome(player: Hand, dealer: Hand) -> EvaluationResult {
        // Player bust always loses regardless of dealer
        if player.isBust {
            return EvaluationResult(outcome: .bust, payoutMultiplier: -1)
        }

        // Both blackjack → push
        if player.isBlackjack && dealer.isBlackjack {
            return EvaluationResult(outcome: .push, payoutMultiplier: 0)
        }

        // Player blackjack only → 3:2
        if player.isBlackjack {
            return EvaluationResult(outcome: .blackjack, payoutMultiplier: 1.5)
        }

        // Dealer blackjack beats any non-BJ hand
        // (European no-hole-card: player loses original bet only; doubles/splits
        // that occurred before dealer BJ is revealed are lost at game-loop level)
        if dealer.isBlackjack {
            return EvaluationResult(outcome: .loss, payoutMultiplier: -1)
        }

        // Dealer bust → player wins
        if dealer.isBust {
            return EvaluationResult(outcome: .win, payoutMultiplier: 1)
        }

        let ps = player.bestScore
        let ds = dealer.bestScore
        if      ps > ds  { return EvaluationResult(outcome: .win,  payoutMultiplier:  1) }
        else if ps < ds  { return EvaluationResult(outcome: .loss, payoutMultiplier: -1) }
        else             { return EvaluationResult(outcome: .push, payoutMultiplier:  0) }
    }

    // MARK: – Perfect Pairs

    static func perfectPairResult(card1: Card, card2: Card) -> PerfectPairResult {
        guard card1.rank == card2.rank else { return .none }
        if card1.suit == card2.suit                   { return .perfectPair  }
        if card1.suit.color == card2.suit.color       { return .colouredPair }
        return .mixedPair
    }

    // MARK: – 21+3

    /// Evaluates 21+3 based on player's two cards + dealer upcard.
    static func twentyOnePlusThreeResult(playerCard1: Card, playerCard2: Card, dealerUp: Card) -> TwentyOnePlusThreeResult {
        let cards   = [playerCard1, playerCard2, dealerUp]
        let suits   = cards.map(\.suit)
        let ranks   = cards.map(\.rank).sorted(by: { $0.rawValue < $1.rawValue })

        let allSameSuit = Set(suits).count == 1
        let allSameRank = Set(ranks.map(\.rawValue)).count == 1
        let isSequential: Bool = {
            let values = ranks.map(\.rawValue)
            return values[2] - values[1] == 1 && values[1] - values[0] == 1
        }()

        if allSameRank && allSameSuit { return .suitedTrips   }
        if allSameRank                { return .threeOfAKind  }
        if isSequential && allSameSuit { return .straightFlush }
        if isSequential               { return .straight      }
        if allSameSuit                { return .flush         }
        return .none
    }

    // MARK: – XP Calculation

    /// XP earned for a hand: base points for outcome, penalised for each strategy error.
    static func xpEarned(strategyErrors: [StrategyError], outcome: HandOutcome) -> Int {
        let base: Int
        switch outcome {
        case .blackjack: base = 50
        case .win:       base = 20
        case .push:      base = 10
        case .loss:      base = 5
        case .bust:      base = 2
        }
        let penalty = strategyErrors.count * 8
        return max(0, base - penalty)
    }
}

File: src/domain/engine/handEvaluator.ts
import { Hand } from '../models/hand'
import { Card, suitColor } from '../models/card'
import { HandOutcome, StrategyError } from '../models/gameState'
import { PerfectPairResult, TwentyOnePlusThreeResult } from '../models/sideBet'

export interface EvaluationResult {
  readonly outcome: HandOutcome
  readonly payoutMultiplier: number  // 1.5=BJ, 1=win, 0=push, -1=loss/bust
}

export const HandEvaluator = {
  outcome(player: Hand, dealer: Hand): EvaluationResult {
    if (player.isBust)
      return { outcome: HandOutcome.BUST, payoutMultiplier: -1 }
    if (player.isBlackjack && dealer.isBlackjack)
      return { outcome: HandOutcome.PUSH, payoutMultiplier: 0 }
    if (player.isBlackjack)
      return { outcome: HandOutcome.BLACKJACK, payoutMultiplier: 1.5 }
    if (dealer.isBlackjack)
      return { outcome: HandOutcome.LOSS, payoutMultiplier: -1 }
    if (dealer.isBust)
      return { outcome: HandOutcome.WIN, payoutMultiplier: 1 }
    const ps = player.bestScore, ds = dealer.bestScore
    if (ps > ds) return { outcome: HandOutcome.WIN,  payoutMultiplier:  1 }
    if (ps < ds) return { outcome: HandOutcome.LOSS, payoutMultiplier: -1 }
    return { outcome: HandOutcome.PUSH, payoutMultiplier: 0 }
  },

  perfectPairResult(card1: Card, card2: Card): PerfectPairResult {
    if (card1.rank !== card2.rank) return PerfectPairResult.NONE
    if (card1.suit === card2.suit)                           return PerfectPairResult.PERFECT_PAIR
    if (suitColor(card1.suit) === suitColor(card2.suit))     return PerfectPairResult.COLOURED_PAIR
    return PerfectPairResult.MIXED_PAIR
  },

  twentyOnePlusThreeResult(p1: Card, p2: Card, dealer: Card): TwentyOnePlusThreeResult {
    const cards  = [p1, p2, dealer]
    const suits  = cards.map(c => c.suit)
    const values = cards.map(c => c.rank as number).sort((a, b) => a - b)

    const allSameSuit = new Set(suits).size === 1
    const allSameRank = new Set(values).size === 1
    const isSequential =
      (values[2]! - values[1]! === 1 && values[1]! - values[0]! === 1) ||
      // Ace-low: ACE.rawValue=14 sorts last, so A-2-3 becomes [2, 3, 14]
      (values[0] === 2 && values[1] === 3 && values[2] === 14)

    if (allSameRank && allSameSuit) return TwentyOnePlusThreeResult.SUITED_TRIPS
    if (allSameRank)                return TwentyOnePlusThreeResult.THREE_OF_KIND
    if (isSequential && allSameSuit) return TwentyOnePlusThreeResult.STRAIGHT_FLUSH
    if (isSequential)               return TwentyOnePlusThreeResult.STRAIGHT
    if (allSameSuit)                return TwentyOnePlusThreeResult.FLUSH
    return TwentyOnePlusThreeResult.NONE
  },

  xpEarned(errors: StrategyError[], outcome: HandOutcome): number {
    const base: Record<HandOutcome, number> = {
      [HandOutcome.BLACKJACK]: 50,
      [HandOutcome.WIN]:       20,
      [HandOutcome.PUSH]:      10,
      [HandOutcome.LOSS]:       5,
      [HandOutcome.BUST]:       2,
    }
    return Math.max(0, base[outcome] - errors.length * 8)
  },
}
